File: frontend/src/mahjong-puzzle/validate.ts
import { Calculator } from '@/mahjong-calc/calc';
import { Rule } from '@/mahjong-calc/definition';
import {
  State,
  PositionType,
  TSUMO,
  RON,
  comparePai,
  type Block,
} from '@/mahjong-calc/types';
import {
  compareMeldFeedback,
  compareOpenHandFeedback,
  isOpenAnswerCorrect,
  isOpenGuessComplete,
  meldsToBlocks,
} from './meld';
import { computeShanten } from './shanten';
import { buildCanonicalAnswer, tileToPai, tilesToC34 } from './tiles';
import type { AgariWay, QueMiOpenGuess, QueMiPuzzle, QueMiOpenSubmitFeedback, TileFeedback, Wind } from './types';

const WIND_TO_POS: Record<Wind, PositionType> = {
  east: PositionType.EAST,
  south: PositionType.SOUTH,
  west: PositionType.WEST,
  north: PositionType.NORTH,
};

export function isCompleteGuess(guess: (string | null)[]): guess is string[] {
  return guess.length === 14 && guess.every((t) => t && t.length > 0);
}

function buildCalcState(
  hand13: string[],
  draw: string,
  field: Wind,
  seat: Wind,
  agariWay: AgariWay,
  dora: string[],
  furu: Block[] = [],
): State {
  const hand = hand13.map(tileToPai).sort(comparePai);
  const agariPai = tileToPai(draw);
  const doraPai = dora.map(tileToPai);
  return new State(
    WIND_TO_POS[field],
    WIND_TO_POS[seat],
    [],
    agariWay === 'tsumo' ? TSUMO : RON,
    hand,
    furu,
    doraPai,
    [],
    agariPai,
    0,
  );
}

const KOKUSHI_INDICES = [0, 8, 9, 17, 18, 26, 27, 28, 29, 30, 31, 32, 33];

/** 14 张均为 7 对（每种 2 枚）。七对子本身 2 翻，荣和/自摸均有役。 */
export function isChiitoitsuShape(tiles14: string[]): boolean {
  if (tiles14.length !== 14) return false;
  let pairs = 0;
  for (const c of tilesToC34(tiles14)) {
    if (c !== 0 && c !== 2) return false;
    if (c === 2) pairs++;
  }
  return pairs === 7;
}

/** 13 种幺九牌各至少 1 枚、其中一种 2 枚，共 14 张。国士无双为役满。 */
export function isKokushiShape(tiles14: string[]): boolean {
  if (tiles14.length !== 14) return false;
  const c34 = tilesToC34(tiles14);
  let pairCount = 0;
  for (const i of KOKUSHI_INDICES) {
    const c = c34[i] ?? 0;
    if (c === 0) return false;
    if (c > 2) return false;
    if (c === 2) pairCount++;
  }
  for (let i = 0; i < 34; i++) {
    if (!KOKUSHI_INDICES.includes(i) && (c34[i] ?? 0) > 0) return false;
  }
  return pairCount === 1;
}

export function isWinningHand(
  hand13: string[],
  draw: string,
  field: Wind,
  seat: Wind,
  agariWay: AgariWay,
  dora: string[],
  furu: Block[] = [],
): boolean {
  const res = new Calculator().calculate(buildCalcState(hand13, draw, field, seat, agariWay, dora, furu), new Rule());
  return res.isYakuman || res.hanRealYaku > 0;
}

export type ValidateResult =
  | { ok: true; correct: boolean }
  | { ok: false; reason: 'incomplete' | 'notWinning' | 'noYaku' | 'isWinning' | 'shantenMismatch' };

/** Non-winnable puzzles must not be ron-able; winnable puzzles use the puzzle's agari way. */
function validationAgariWay(puzzle: QueMiPuzzle): AgariWay {
  return puzzle.type === 'nonWinnable' ? 'ron' : puzzle.agariWay;
}

function validateSubmitAgari(
  puzzle: QueMiPuzzle,
  hand13: string[],
  draw: string,
  furu: Block[] = [],
): ValidateResult | null {
  const required = validationAgariWay(puzzle);
  const wins = isWinningHand(hand13, draw, puzzle.fieldWind, puzzle.seatWind, required, puzzle.dora, furu);

  if (puzzle.type === 'winnable') {
    if (wins) return null;
    const alt: AgariWay = required === 'ron' ? 'tsumo' : 'ron';
    if (isWinningHand(hand13, draw, puzzle.fieldWind, puzzle.seatWind, alt, puzzle.dora, furu)) {
      return { ok: false, reason: 'noYaku' };
    }
    return { ok: false, reason: 'notWinning' };
  }

  if (isWinningHand(hand13, draw, puzzle.fieldWind, puzzle.seatWind, 'ron', puzzle.dora, furu)) {
    return { ok: false, reason: 'isWinning' };
  }
  return null;
}

export function getAnswerYaku(puzzle: QueMiPuzzle): string[] {
  if (puzzle.handMode === 'open' && puzzle.openAnswer) {
    const { closedHand, draw, melds } = puzzle.openAnswer;
    const furu = meldsToBlocks(melds);
    const res = new Calculator().calculate(
      buildCalcState(closedHand, draw, puzzle.fieldWind, puzzle.seatWind, puzzle.agariWay, puzzle.dora, furu),
      new Rule(),
    );
    return res.yaku;
  }
  const hand13 = puzzle.answer.slice(0, 13);
  const draw = puzzle.answer[13]!;
  const res = new Calculator().calculate(
    buildCalcState(hand13, draw, puzzle.fieldWind, puzzle.seatWind, puzzle.agariWay, puzzle.dora),
    new Rule(),
  );
  return res.yaku;
}

/** 提示用役种（不含宝牌） */
export function getAnswerYakuHint(puzzle: QueMiPuzzle): string[] {
  return getAnswerYaku(puzzle).filter((y) => !y.includes('宝牌'));
}

export function isKokushiWin(
  hand13: string[],
  draw: string,
  field: Wind,
  seat: Wind,
  agariWay: AgariWay,
  dora: string[],
  furu: Block[] = [],
): boolean {
  const res = new Calculator().calculate(buildCalcState(hand13, draw, field, seat, agariWay, dora, furu), new Rule());
  return res.isYakuman && res.yaku.some((y) => y.includes('国士'));
}

export function validateGuess(puzzle: QueMiPuzzle, guess: (string | null)[]): ValidateResult {
  if (puzzle.handMode === 'open') {
    return { ok: false, reason: 'incomplete' };
  }
  if (!isCompleteGuess(guess)) return { ok: false, reason: 'incomplete' };

  const hand13 = guess.slice(0, 13);
  const draw = guess[13]!;
  const agariErr = validateSubmitAgari(puzzle, hand13, draw);
  if (agariErr) return agariErr;

  if (puzzle.type === 'nonWinnable') {
    if (puzzle.shanten != null && computeShanten(hand13) !== puzzle.shanten) {
      return { ok: false, reason: 'shantenMismatch' };
    }
  }

  const correct = guess.every((t, i) => t === puzzle.answer[i]);
  return { ok: true, correct };
}

export function validateOpenGuess(puzzle: QueMiPuzzle, openGuess: QueMiOpenGuess): ValidateResult {
  if (puzzle.handMode !== 'open' || puzzle.openMeldCount == null) {
    return { ok: false, reason: 'incomplete' };
  }
  if (!isOpenGuessComplete(puzzle.openMeldCount, openGuess.melds, openGuess.hand)) {
    return { ok: false, reason: 'incomplete' };
  }

  const guessMelds = openGuess.melds as string[][];
  const guessHand = openGuess.hand as string[];
  const furu = meldsToBlocks(guessMelds);
  const draw = guessHand[guessHand.length - 1]!;
  const hand13 = guessHand.slice(0, -1) as string[];

  const agariErr = validateSubmitAgari(puzzle, hand13, draw, furu);
  if (agariErr) return agariErr;

  if (puzzle.type === 'nonWinnable') {
    if (puzzle.shanten != null && computeShanten(hand13) !== puzzle.shanten) {
      return { ok: false, reason: 'shantenMismatch' };
    }
  }

  const correct = puzzle.openAnswer
    ? isOpenAnswerCorrect(puzzle.openAnswer, openGuess.melds, openGuess.hand)
    : false;
  return { ok: true, correct };
}

export function compareGuessFeedback(answer: string[], guess: string[]): TileFeedback[] {
  const feedback: TileFeedback[] = new Array(14).fill('black');
  const answerRemaining = countMultiset(answer);
  const guessRemaining = countMultiset(guess);

  for (let i = 0; i < 14; i++) {
    if (guess[i] === answer[i]) {
      feedback[i] = 'green';
      answerRemaining[guess[i]!] = (answerRemaining[guess[i]!] ?? 0) - 1;
      guessRemaining[guess[i]!] = (guessRemaining[guess[i]!] ?? 0) - 1;
    }
  }

  for (let i = 0; i < 14; i++) {
    if (feedback[i] === 'green') continue;
    const t = guess[i];
    if (!t) {
      feedback[i] = 'black';
      continue;
    }
    if ((answerRemaining[t] ?? 0) > 0) {
      feedback[i] = 'yellow';
      answerRemaining[t]!--;
    } else {
      feedback[i] = 'black';
    }
  }

  return feedback;
}

export function compareOpenGuessFeedback(puzzle: QueMiPuzzle, openGuess: QueMiOpenGuess): QueMiOpenSubmitFeedback {
  const answer = puzzle.openAnswer!;
  return {
    meldFeedback: compareMeldFeedback(answer.melds, openGuess.melds),
    handFeedback: compareOpenHandFeedback(answer, openGuess.hand),
  };
}

function countMultiset(tiles: string[]): Record<string, number> {
  const c: Record<string, number> = {};
  for (const t of tiles) c[t] = (c[t] ?? 0) + 1;
  return c;
}

export function formatAnswerTiles(hand13: string[], draw: string): string[] {
  return buildCanonicalAnswer(hand13, draw);
}

File: backend/quemi/validate.go
package quemi

import (
	"strings"
)

var windToPos = map[Wind]PositionType{
	WindEast:  PositionEast,
	WindSouth: PositionSouth,
	WindWest:  PositionWest,
	WindNorth: PositionNorth,
}

// ValidateReason describes why validation failed.
type ValidateReason string

const (
	ReasonIncomplete       ValidateReason = "incomplete"
	ReasonNotWinning       ValidateReason = "notWinning"
	ReasonNoYaku           ValidateReason = "noYaku"
	ReasonIsWinning        ValidateReason = "isWinning"
	ReasonShantenMismatch  ValidateReason = "shantenMismatch"
	ReasonInvalidTiles     ValidateReason = "invalidTiles"
	ReasonDuplicateTiles   ValidateReason = "duplicateTiles"
	ReasonKokushi          ValidateReason = "kokushi"
	ReasonInvalidStructure ValidateReason = "invalidStructure"
)

// ValidateResult is the outcome of guess or definition validation.
type ValidateResult struct {
	OK      bool
	Correct bool
	Reason  ValidateReason
}

// IsCompleteGuess returns true when all 14 closed slots are filled.
func IsCompleteGuess(guess []string) bool {
	if len(guess) != 14 {
		return false
	}
	for _, t := range guess {
		if t == "" {
			return false
		}
	}
	return true
}

func buildCalcState(hand13 []string, draw string, field, seat Wind, agariWay AgariWay, dora []string, furu []Block) (State, error) {
	hand := make([]Pai, len(hand13))
	for i, t := range hand13 {
		p, err := TileToPai(t)
		if err != nil {
			return State{}, err
		}
		hand[i] = p
	}
	sortPaiSlice(hand)
	agariPai, err := TileToPai(draw)
	if err != nil {
		return State{}, err
	}
	doraPai := make([]Pai, len(dora))
	for i, t := range dora {
		doraPai[i] = MustTileToPai(t)
	}
	agariFlag := Tsumo
	if agariWay == AgariWayRon {
		agariFlag = Ron
	}
	return NewState(windToPos[field], windToPos[seat], nil, agariFlag, hand, furu, doraPai, nil, agariPai, 0), nil
}

func sortPaiSlice(pais []Pai) {
	for i := 0; i < len(pais); i++ {
		for j := i + 1; j < len(pais); j++ {
			if ComparePai(pais[j], pais[i]) < 0 {
				pais[i], pais[j] = pais[j], pais[i]
			}
		}
	}
}

func calcResult(hand13 []string, draw string, field, seat Wind, agariWay AgariWay, dora []string, furu []Block) (Result, error) {
	st, err := buildCalcState(hand13, draw, field, seat, agariWay, dora, furu)
	if err != nil {
		return Result{}, err
	}
	return NewCalculator().Calculate(st), nil
}

// IsWinningHand returns true when the hand wins with yaku or yakuman.
func IsWinningHand(hand13 []string, draw string, field, seat Wind, agariWay AgariWay, dora []string, furu []Block) bool {
	res, err := calcResult(hand13, draw, field, seat, agariWay, dora, furu)
	if err != nil {
		return false
	}
	return res.IsYakuman || res.HanRealYaku > 0
}

// IsKokushiWin returns true when the hand is a kokushi yakuman.
func IsKokushiWin(hand13 []string, draw string, field, seat Wind, agariWay AgariWay, dora []string, furu []Block) bool {
	res, err := calcResult(hand13, draw, field, seat, agariWay, dora, furu)
	if err != nil {
		return false
	}
	if !res.IsYakuman {
		return false
	}
	for _, y := range res.Yaku {
		if strings.Contains(y, "国士") {
			return true
		}
	}
	return false
}

// ValidateGuess validates a closed-hand player guess.
func ValidateGuess(puzzle QueMiPuzzle, guess []string) ValidateResult {
	if puzzle.HandMode == HandModeOpen {
		return ValidateResult{OK: false, Reason: ReasonIncomplete}
	}
	if !IsCompleteGuess(guess) {
		return ValidateResult{OK: false, Reason: ReasonIncomplete}
	}
	hand13 := guess[:13]
	draw := guess[13]
	winning := IsWinningHand(hand13, draw, puzzle.FieldWind, puzzle.SeatWind, puzzle.AgariWay, puzzle.Dora, nil)

	if puzzle.Type == PuzzleTypeWinnable {
		if !winning {
			return ValidateResult{OK: false, Reason: ReasonNotWinning}
		}
	} else {
		if winning {
			return ValidateResult{OK: false, Reason: ReasonIsWinning}
		}
		if puzzle.Shanten != nil && ComputeShanten(hand13) != *puzzle.Shanten {
			return ValidateResult{OK: false, Reason: ReasonShantenMismatch}
		}
	}

	correct := true
	for i, t := range guess {
		if i >= len(puzzle.Answer) || t != puzzle.Answer[i] {
			correct = false
			break
		}
	}
	return ValidateResult{OK: true, Correct: correct}
}

// ValidateOpenGuess validates an open-hand player guess.
func ValidateOpenGuess(puzzle QueMiPuzzle, openGuess QueMiOpenGuess) ValidateResult {
	if puzzle.HandMode != HandModeOpen || puzzle.OpenAnswer == nil || puzzle.OpenMeldCount == nil {
		return ValidateResult{OK: false, Reason: ReasonIncomplete}
	}
	if !IsOpenGuessComplete(*puzzle.OpenMeldCount, openGuess.Melds, openGuess.Hand) {
		return ValidateResult{OK: false, Reason: ReasonIncomplete}
	}

	furu := MeldsToBlocks(openGuess.Melds)
	draw := openGuess.Hand[len(openGuess.Hand)-1]
	hand13 := openGuess.Hand[:len(openGuess.Hand)-1]

	winCheck := IsWinningHand(hand13, draw, puzzle.FieldWind, puzzle.SeatWind, puzzle.AgariWay, puzzle.Dora, furu)
	if puzzle.Type == PuzzleTypeWinnable && !winCheck {
		return ValidateResult{OK: false, Reason: ReasonNotWinning}
	}

	correct := IsOpenAnswerCorrect(*puzzle.OpenAnswer, openGuess.Melds, openGuess.Hand)
	return ValidateResult{OK: true, Correct: correct}
}

// CompareGuessFeedback returns Wordle-style feedback for closed guesses.
func CompareGuessFeedback(answer, guess []string) []TileFeedback {
	feedback := make([]TileFeedback, 14)
	for i := range feedback {
		feedback[i] = FeedbackBlack
	}
	answerRemaining := CountTiles(answer)
	guessRemaining := CountTiles(guess)

	for i := 0; i < 14 && i < len(guess) && i < len(answer); i++ {
		if guess[i] == answer[i] {
			feedback[i] = FeedbackGreen
			answerRemaining[guess[i]]--
			guessRemaining[guess[i]]--
		}
	}
	for i := 0; i < 14 && i < len(guess); i++ {
		if feedback[i] == FeedbackGreen {
			continue
		}
		t := guess[i]
		if t == "" {
			feedback[i] = FeedbackBlack
			continue
		}
		if answerRemaining[t] > 0 {
			feedback[i] = FeedbackYellow
			answerRemaining[t]--
		} else {
			feedback[i] = FeedbackBlack
		}
	}
	return feedback
}

// CompareOpenGuessFeedback returns feedback for open guesses.
func CompareOpenGuessFeedback(puzzle QueMiPuzzle, openGuess QueMiOpenGuess) QueMiOpenSubmitFeedback {
	answer := puzzle.OpenAnswer
	return QueMiOpenSubmitFeedback{
		MeldFeedback: CompareMeldFeedback(answer.Melds, openGuess.Melds),
		HandFeedback: CompareOpenHandFeedback(*answer, openGuess.Hand),
	}
}

func puzzleHandAndFuru(puzzle QueMiPuzzle) (hand13 []string, draw string, furu []Block, err error) {
	if puzzle.HandMode == HandModeOpen && puzzle.OpenAnswer != nil {
		oa := puzzle.OpenAnswer
		furu = MeldsToBlocks(oa.Melds)
		return oa.ClosedHand, oa.Draw, furu, nil
	}
	if len(puzzle.Answer) != 14 {
		return nil, "", nil, errInvalidStructure
	}
	return puzzle.Answer[:13], puzzle.Answer[13], nil, nil
}

var errInvalidStructure = validateError("invalid structure")

type validateError string

func (e validateError) Error() string { return string(e) }

func countAllPuzzleTiles(puzzle QueMiPuzzle) map[string]int {
	c := CountTiles(puzzle.Dora)
	if puzzle.HandMode == HandModeOpen && puzzle.OpenAnswer != nil {
		c = mergeCounts(c, CountTiles(flattenMelds(puzzle.OpenAnswer.Melds)))
		c = mergeCounts(c, CountTiles(puzzle.OpenAnswer.ClosedHand))
		if puzzle.OpenAnswer.Draw != "" {
			c[puzzle.OpenAnswer.Draw]++
		}
	} else if len(puzzle.Answer) == 14 {
		c = mergeCounts(c, CountTiles(puzzle.Answer))
	}
	return c
}

func mergeCounts(a, b map[string]int) map[string]int {
	out := make(map[string]int)
	for k, v := range a {
		out[k] = v
	}
	for k, v := range b {
		out[k] += v
	}
	return out
}

func tilesValid(tiles map[string]int) bool {
	for t, n := range tiles {
		if TileToIndex(t) < 0 {
			return false
		}
		if n > 4 {
			return false
		}
	}
	return true
}

// ValidatePuzzleDefinition checks puzzle authoring rules.
func ValidatePuzzleDefinition(puzzle QueMiPuzzle) ValidateResult {
	if puzzle.HandMode == HandModeOpen {
		if puzzle.OpenAnswer == nil || puzzle.OpenMeldCount == nil {
			return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
		}
		if len(puzzle.OpenAnswer.Melds) != *puzzle.OpenMeldCount {
			return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
		}
	} else if len(puzzle.Answer) != 14 {
		return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
	}

	allTiles := countAllPuzzleTiles(puzzle)
	if !tilesValid(allTiles) {
		for t, n := range allTiles {
			if TileToIndex(t) < 0 {
				_ = t
				return ValidateResult{OK: false, Reason: ReasonInvalidTiles}
			}
			if n > 4 {
				return ValidateResult{OK: false, Reason: ReasonDuplicateTiles}
			}
		}
	}

	hand13, draw, furu, err := puzzleHandAndFuru(puzzle)
	if err != nil {
		return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
	}

	winning := IsWinningHand(hand13, draw, puzzle.FieldWind, puzzle.SeatWind, puzzle.AgariWay, puzzle.Dora, furu)
	kokushi := IsKokushiWin(hand13, draw, puzzle.FieldWind, puzzle.SeatWind, puzzle.AgariWay, puzzle.Dora, furu)

	switch puzzle.Type {
	case PuzzleTypeWinnable:
		if !winning {
			return ValidateResult{OK: false, Reason: ReasonNotWinning}
		}
		if kokushi {
			return ValidateResult{OK: false, Reason: ReasonKokushi}
		}
	case PuzzleTypeNonWinnable:
		if winning {
			return ValidateResult{OK: false, Reason: ReasonIsWinning}
		}
		if puzzle.Shanten == nil {
			return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
		}
		if ComputeShanten(hand13) != *puzzle.Shanten {
			return ValidateResult{OK: false, Reason: ReasonShantenMismatch}
		}
	default:
		return ValidateResult{OK: false, Reason: ReasonInvalidStructure}
	}

	return ValidateResult{OK: true, Correct: true}
}
